src/control/server/storage/bdev.go

```diff
@@ -58,56 +58,102 @@ const (
 	AccelOptCRCFlag  = C.NVME_ACCEL_FLAG_CRC
 )
 
-// NvmeDevState represents the health state of NVMe device as reported by DAOS engine BIO module.
-type NvmeDevState uint32
+// NvmeDevState represents the operation state of an NVMe device.
+type NvmeDevState int32
 
-// NvmeDevState values representing individual bit-flags.
+// NvmeDevState values representing the operational device state.
 const (
-	NvmeStatePlugged  NvmeDevState = C.NVME_DEV_FL_PLUGGED
-	NvmeStateInUse    NvmeDevState = C.NVME_DEV_FL_INUSE
-	NvmeStateFaulty   NvmeDevState = C.NVME_DEV_FL_FAULTY
-	NvmeStateIdentify NvmeDevState = C.NVME_DEV_FL_IDENTIFY
-	NvmeStateUnknown  NvmeDevState = C.NVME_DEV_STATE_INVALID
+	NvmeStateNormal NvmeDevState = iota
+	NvmeStateNew
+	NvmeStateFaulty
 )
 
-// IsNew returns true if SSD is not in use by DAOS.
-func (bs NvmeDevState) IsNew() bool {
-	return bs&NvmeStatePlugged != 0 && bs&NvmeStateInUse == 0
+func (nds NvmeDevState) String() string {
+	ndss, ok := ctlpb.NvmeDevState_name[int32(nds)]
+	if !ok {
+		return "UNKNOWN"
+	}
+	return strings.ToUpper(ndss)
 }
 
-// IsNormal returns true if SSD is in a normal, non-faulty state.
-func (bs NvmeDevState) IsNormal() bool {
-	return bs&NvmeStatePlugged != 0 && bs&NvmeStateFaulty == 0 && bs&NvmeStateInUse != 0
+func (nds NvmeDevState) MarshalJSON() ([]byte, error) {
+	stateStr, ok := ctlpb.NvmeDevState_name[int32(nds)]
+	if !ok {
+		return nil, errors.Errorf("invalid nvme dev state %d", nds)
+	}
+	return []byte(`"` + strings.ToUpper(stateStr) + `"`), nil
 }
 
-// IsFaulty returns true if SSD is in a faulty state.
-func (bs NvmeDevState) IsFaulty() bool {
-	return bs&NvmeStatePlugged != 0 && bs&NvmeStateFaulty != 0
-}
+func (nds *NvmeDevState) UnmarshalJSON(data []byte) error {
+	stateStr := strings.Trim(strings.ToUpper(string(data)), "\"")
+
+	state, ok := ctlpb.NvmeDevState_value[stateStr]
+	if !ok {
+		// Try converting the string to an int32, to handle the
+		// conversion from protobuf message using convert.Types().
+		si, err := strconv.ParseInt(stateStr, 0, 32)
+		if err != nil {
+			return errors.Errorf("invalid nvme dev state number parse %q", stateStr)
+		}
+
+		if _, ok = ctlpb.NvmeDevState_name[int32(si)]; !ok {
+			return errors.Errorf("invalid nvme dev state name lookup %q", stateStr)
+		}
+		state = int32(si)
+	}
+	*nds = NvmeDevState(state)
 
-func (bs NvmeDevState) String() string {
-	return C.GoString(C.nvme_state2str(C.int(bs)))
+	return nil
 }
 
-// States lists each flag in state bitset.
-func (bs NvmeDevState) States() string {
-	return fmt.Sprintf("plugged: %v, in-use: %v, faulty: %v, identify: %v",
-		bs&C.NVME_DEV_FL_PLUGGED != 0, bs&C.NVME_DEV_FL_INUSE != 0,
-		bs&C.NVME_DEV_FL_FAULTY != 0, bs&C.NVME_DEV_FL_IDENTIFY != 0)
+// VmdLedState represents the LED state of VMD device.
+type VmdLedState int32
```

Comment:
nit: Is there a reason to make this type specifically about VMD LEDs? What if we add support for SCSI/SAS/SATA LEDs? Or some other type of device? As with the previous comment, I think it would be better to keep VMD contained as an implementation detail rather than exposing it in these exported types.

Reply:
done